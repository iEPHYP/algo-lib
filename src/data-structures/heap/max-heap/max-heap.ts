import { AbstractHeap } from '../abstract-heap';

export class MaxHeap<
  TItem extends number | Record<string, any> = number,
> extends AbstractHeap<TItem> {
  // Complexity O(logN)
  public insert(newItem: TItem) {
    this.items.push(newItem);
    const newItemValue = this.getValue(newItem);

    let newItemIndex = this.items.length - 1;
    let parentIndex = this.getParentIndex(newItemIndex);
    let parentValue = this.getValueAt(parentIndex);
    while (parentValue !== undefined && newItemValue > parentValue) {
      this.swapItems(parentIndex, newItemIndex);

      newItemIndex = parentIndex;
      parentIndex = this.getParentIndex(newItemIndex);
      parentValue = this.getValueAt(parentIndex);
    }
  }

  // Complexity O(logN)
  public extractMax() {
    const root = this.items[this.firstIndex()];

    if (this.size() <= 1) {
      return this.items.pop();
    }

    this.swapItems(this.firstIndex(), this.lastIndex());

    this.items.pop();

    // TODO: make this less verbose
    const sinkDown = (unstablizedNewRootIndex: number) => {
      const unstablizedNewRootValue = this.getValueAt(unstablizedNewRootIndex);

      if (unstablizedNewRootValue === undefined) {
        return;
      }

      const leftChildIndex = this.getLeftChildIndex(unstablizedNewRootIndex);
      const leftChildValue = this.getValueAt(leftChildIndex);

      const rightChildIndex = this.getRightChildIndex(unstablizedNewRootIndex);
      const rightChildValue = this.getValueAt(rightChildIndex);

      const swapWithLeftChild = (leftChildValue: number) => {
        if (leftChildValue > unstablizedNewRootValue) {
          this.swapItems(leftChildIndex, unstablizedNewRootIndex);
          sinkDown(leftChildIndex);
        }
      };

      const swapWithRightChild = (rightChildValue: number) => {
        if (rightChildValue > unstablizedNewRootValue) {
          this.swapItems(rightChildIndex, unstablizedNewRootIndex);
          sinkDown(rightChildIndex);
        }
      };

      if (leftChildValue !== undefined && rightChildValue !== undefined) {
        if (leftChildValue > rightChildValue) {
          swapWithLeftChild(leftChildValue);
        } else {
          swapWithRightChild(rightChildValue);
        }
      } else {
        if (leftChildValue !== undefined) {
          swapWithLeftChild(leftChildValue);
        } else if (rightChildValue !== undefined) {
          swapWithRightChild(rightChildValue);
        }
      }
    };

    sinkDown(this.firstIndex());

    return root;
  }
}
