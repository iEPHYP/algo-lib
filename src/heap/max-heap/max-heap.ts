// TODO: realize MinHeap
export class MaxHeap<TItem extends { value: number } | number = number> {
  private items: (TItem | undefined)[] = [];
  private getValue: (item: TItem) => number = (item) =>
    typeof item === 'number' ? item : item.value;

  private getValueAt(index: number) {
    const item = this.items[index];

    return item !== undefined ? this.getValue(item) : undefined;
  }

  constructor(getValue?: (item: TItem) => number) {
    this.getValue = getValue ?? this.getValue;
  }

  private getLeftChildIndex(parentIndex: number) {
    return 2 * parentIndex + 1;
  }

  private getLeftChild(parentIndex: number) {
    return this.items[this.getLeftChildIndex(parentIndex)];
  }

  private getRightChildIndex(parentIndex: number) {
    return 2 * parentIndex + 2;
  }

  private getRightChild(parentIndex: number) {
    return this.items[this.getRightChildIndex(parentIndex)];
  }

  private getParentIndex(childIndex: number) {
    return Math.floor((childIndex - 1) / 2);
  }

  private getParent(childIndex: number) {
    return this.items[this.getParentIndex(childIndex)];
  }

  private swapItems(item1Index: number, item2Index: number) {
    const item1 = this.items[item1Index];
    const item2 = this.items[item2Index];
    this.items[item1Index] = item2;
    this.items[item2Index] = item1;
  }

  // Complexity O(logN) AVG, O(N) WORST
  public insert(newItem: TItem) {
    this.items.push(newItem);
    const newItemValue = this.getValue(newItem);

    let newItemIndex = this.items.length - 1;
    let parentIndex = this.getParentIndex(newItemIndex);
    let parentValue = this.getValueAt(parentIndex);
    while (parentValue !== undefined && newItemValue > parentValue) {
      this.swapItems(parentIndex, newItemIndex);

      newItemIndex = parentIndex;
      parentIndex = this.getParentIndex(newItemIndex);
      parentValue = this.getValueAt(parentIndex);
    }
  }

  // Complexity O(logN) AVG, O(N) WORST
  public extractMax() {
    const root = this.items[this.firstIndex()];

    if (this.size() <= 1) {
      return this.items.pop();
    }

    this.swapItems(this.firstIndex(), this.lastIndex());

    this.items.pop();

    // TODO: make this less verbose
    const sinkDown = (unstablizedNewRootIndex: number) => {
      const unstablizedNewRootValue = this.getValueAt(unstablizedNewRootIndex);

      if (unstablizedNewRootValue === undefined) {
        return;
      }

      const leftChildIndex = this.getLeftChildIndex(unstablizedNewRootIndex);
      const leftChildValue = this.getValueAt(leftChildIndex);

      const rightChildIndex = this.getRightChildIndex(unstablizedNewRootIndex);
      const rightChildValue = this.getValueAt(rightChildIndex);

      const swapWithLeftChild = (leftChildValue: number) => {
        if (leftChildValue > unstablizedNewRootValue) {
          this.swapItems(leftChildIndex, unstablizedNewRootIndex);
          sinkDown(leftChildIndex);
        }
      };

      const swapWithRightChild = (rightChildValue: number) => {
        if (rightChildValue > unstablizedNewRootValue) {
          this.swapItems(rightChildIndex, unstablizedNewRootIndex);
          sinkDown(rightChildIndex);
        }
      };

      if (leftChildValue !== undefined && rightChildValue !== undefined) {
        if (leftChildValue > rightChildValue) {
          swapWithLeftChild(leftChildValue);
        } else {
          swapWithRightChild(rightChildValue);
        }
      } else {
        if (leftChildValue !== undefined) {
          swapWithLeftChild(leftChildValue);
        } else if (rightChildValue !== undefined) {
          swapWithRightChild(rightChildValue);
        }
      }
    };

    sinkDown(this.firstIndex());

    return root;
  }

  public size() {
    return this.items.length;
  }

  private firstIndex() {
    return 0;
  }

  private lastIndex() {
    return this.items.length - 1;
  }

  public toList() {
    return this.items;
  }

  // TODO: realize this method and use for Snapshots
  // public toString() {}
}
