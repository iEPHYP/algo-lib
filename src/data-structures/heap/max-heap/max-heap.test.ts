import { MaxHeap } from './max-heap';

describe('MaxHeap', () => {
  let maxHeap = new MaxHeap();

  beforeEach(() => {
    maxHeap = new MaxHeap();
  });

  it('inserts items properly', () => {
    maxHeap.insert(1);
    expect(maxHeap.toList()).toMatchSnapshot();
    maxHeap.insert(5);
    expect(maxHeap.toList()).toMatchSnapshot();
    maxHeap.insert(3);
    expect(maxHeap.toList()).toMatchSnapshot();
    maxHeap.insert(4);
    expect(maxHeap.toList()).toMatchSnapshot();
    maxHeap.insert(2);
    expect(maxHeap.toList()).toMatchSnapshot();
    maxHeap.insert(0);
    expect(maxHeap.toList()).toMatchSnapshot();
    maxHeap.insert(10);
    expect(maxHeap.toList()).toMatchSnapshot();
  });

  it('extracts Max root properly', () => {
    maxHeap.insert(1);
    expect(maxHeap.extractMax()).toBe(1);
    expect(maxHeap.toList()).toMatchSnapshot('extracting from heap with 1 item');
    maxHeap.insert(1);
    maxHeap.insert(5);
    expect(maxHeap.extractMax()).toBe(5);
    expect(maxHeap.toList()).toMatchSnapshot('extracting from heap with 2 items');
    maxHeap.insert(5);
    maxHeap.insert(0);
    maxHeap.insert(4);
    maxHeap.insert(2);
    maxHeap.insert(3);
    maxHeap.insert(10);
    maxHeap.insert(6);
    maxHeap.insert(19);
    expect(maxHeap.toList()).toMatchSnapshot('heap with several items');
    expect(maxHeap.extractMax()).toBe(19);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 19 from heap with several items');
    expect(maxHeap.extractMax()).toBe(10);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 10 from heap with several items');
    expect(maxHeap.extractMax()).toBe(6);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 6 from heap with several items');
    expect(maxHeap.extractMax()).toBe(5);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 5 from heap with several items');
    expect(maxHeap.extractMax()).toBe(4);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 4 from heap with several items');
    expect(maxHeap.extractMax()).toBe(3);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 3 from heap with several items');
    expect(maxHeap.extractMax()).toBe(2);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 2 from heap with several items');
    expect(maxHeap.extractMax()).toBe(1);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 1 from heap with several items');
    expect(maxHeap.extractMax()).toBe(0);
    expect(maxHeap.toList()).toMatchSnapshot('extracting 0 from heap with several items');
  });
});
