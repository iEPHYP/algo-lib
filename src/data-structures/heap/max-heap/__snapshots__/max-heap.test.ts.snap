// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxHeap extracts Max root properly: extracting 0 from heap with several items 1`] = `[]`;

exports[`MaxHeap extracts Max root properly: extracting 1 from heap with several items 1`] = `
[
  0,
]
`;

exports[`MaxHeap extracts Max root properly: extracting 2 from heap with several items 1`] = `
[
  1,
  0,
]
`;

exports[`MaxHeap extracts Max root properly: extracting 3 from heap with several items 1`] = `
[
  2,
  1,
  0,
]
`;

exports[`MaxHeap extracts Max root properly: extracting 4 from heap with several items 1`] = `
[
  3,
  2,
  0,
  1,
]
`;

exports[`MaxHeap extracts Max root properly: extracting 5 from heap with several items 1`] = `
[
  4,
  2,
  3,
  1,
  0,
]
`;

exports[`MaxHeap extracts Max root properly: extracting 6 from heap with several items 1`] = `
[
  5,
  4,
  3,
  1,
  2,
  0,
]
`;

exports[`MaxHeap extracts Max root properly: extracting 10 from heap with several items 1`] = `
[
  6,
  4,
  5,
  1,
  2,
  0,
  3,
]
`;

exports[`MaxHeap extracts Max root properly: extracting 19 from heap with several items 1`] = `
[
  10,
  6,
  5,
  4,
  2,
  0,
  3,
  1,
]
`;

exports[`MaxHeap extracts Max root properly: extracting from heap with 1 item 1`] = `[]`;

exports[`MaxHeap extracts Max root properly: extracting from heap with 2 items 1`] = `
[
  1,
]
`;

exports[`MaxHeap extracts Max root properly: heap with several items 1`] = `
[
  19,
  10,
  5,
  6,
  2,
  0,
  3,
  1,
  4,
]
`;

exports[`MaxHeap inserts items properly 1`] = `
[
  1,
]
`;

exports[`MaxHeap inserts items properly 2`] = `
[
  5,
  1,
]
`;

exports[`MaxHeap inserts items properly 3`] = `
[
  5,
  1,
  3,
]
`;

exports[`MaxHeap inserts items properly 4`] = `
[
  5,
  4,
  3,
  1,
]
`;

exports[`MaxHeap inserts items properly 5`] = `
[
  5,
  4,
  3,
  1,
  2,
]
`;

exports[`MaxHeap inserts items properly 6`] = `
[
  5,
  4,
  3,
  1,
  2,
  0,
]
`;

exports[`MaxHeap inserts items properly 7`] = `
[
  10,
  4,
  5,
  1,
  2,
  0,
  3,
]
`;
